### PR TITLE
Timestamp-based versioning for InnoDB [closes #209]

### LINKED ISSUE #209: Timestamp-based versioning for InnoDB
##### InnoDB tables should:
1. have possibility to be versioned on SQL level, like MyISAM tables (this issue);
2. mimic timestamp-based versioning, working internally on TRX_ID basis (not subjected in current issue).

### PR DESCRIPTION
**Pull request by midenok**
Please, review.

### CODE REVIEW

**Comment by kevgs** [comment]
`versioning.select_sp` fails.

**Comment by kevgs** [comment]
After some discussion it was decided to use `system-versioned` instead of `temporal`.

**Comment by kevgs** [comment]
`partition` test works only in `innodb` mode. Should be also `myisam`?
```
versioning.partition 'innodb'            w1 [ pass ]   2137
```

File: storage/innobase/trx/trx0trx.cc [hunk 1]
```diff
@@ -64,9 +64,15 @@ int thd_deadlock_victim_preference(const MYSQL_THD thd1, const MYSQL_THD thd2);
 
 /** The bit pattern corresponding to TRX_ID_MAX */
 const char trx_id_max_bytes[8] = {
-	'\377', '\377', '\377', '\377', '\377', '\377', '\377', '\377'
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
 };
 
+/** The bit pattern corresponding to max timestamp */
+const char timestamp_max_bytes[7] = {
+	0x7f, 0xff, 0xff, 0xff, 0x0f, 0x42, 0x3f
```
Comment (kevgs) [inline storage/innobase/trx/trx0trx.cc:72]: Nothing to fix right now, but this could not work on big-endian architectures. Buildbots will show.

**Comment by kevgs** [comment]
```sql
create or replace table t (a int) with system versioning engine=innodb;
insert into t values (1);
alter table t drop system versioning, algorithm=inplace;
```
```c++
#7  __GI___assert_fail (assertion=0x334135 "len == 8", file=0x373a59 "/home/kevg/work/mariadb/storage/innobase/row/row0merge.cc", line=2330, function=0x239e27 "dberr_t row_merge_read_clustered_index(trx_t *, struct TABLE *, const dict_table_t *, const dict_table_t *, bool, dict_index_t **, dict_index_t *, fts_psort_t *, merge_file_t *, const ulint *, ulint, "...) at assert.c:101
#8  row_merge_read_clustered_index (trx=0x7f45e69220f8, table=0x7f459809f050, old_table=0x7f459816d2a8, new_table=0x7f45980914f8, online=false, index=0x7f4598091268, fts_sort_idx=0x0, psort_info=0x0, files=0x7f45980ad5b0, key_numbers=0x7f4598091270, n_index=1, add_cols=0x0, add_v=0x0, col_map=0x7f45980912b8, add_autoinc=18446744073709551615, sequence=..., block=0x7f45dc2aa000 <error: Cannot access memory at address 0x7f45dc2aa000>, skip_pk_sort=true, tmpfd=0x7f45dc5eb578, stage=0x7f4598028c80, pct_cost=50, crypt_block=0x0, eval_table=0x7f459809f050, drop_historical=false) at row0merge.cc:2330
#9  row_merge_build_indexes (trx=0x7f45e69220f8, old_table=0x7f459816d2a8, new_table=0x7f45980914f8, online=false, indexes=0x7f4598091268, key_numbers=0x7f4598091270, n_indexes=1, table=0x7f459809f050, add_cols=0x0, col_map=0x7f45980912b8, add_autoinc=18446744073709551615, sequence=..., skip_pk_sort=true, stage=0x7f4598028c80, add_v=0x0, eval_table=0x7f459809f050, drop_historical=false) at row0merge.cc:4780
#10 ha_innobase::inplace_alter_table (this=0x7f45980af758, altered_table=0x7f459809f050, ha_alter_info=0x7f45dc5ed500) at handler0alter.cc:7088
#11 handler::ha_inplace_alter_table (this=0x7f45980af758, altered_table=0x7f459809f050, ha_alter_info=0x7f45dc5ed500) at handler.h:4110
#12 mysql_inplace_alter_table (thd=0x7f4598000d50, table_list=0x7f4598015070, table=0x7f45980aeb10, altered_table=0x7f459809f050, ha_alter_info=0x7f45dc5ed500, inplace_supported=HA_ALTER_INPLACE_NO_LOCK_AFTER_PREPARE, target_mdl_request=0x7f45dc5edba0, alter_ctx=0x7f45dc5edd50) at sql_table.cc:7573
#13 mysql_alter_table (thd=0x7f4598000d50, new_db=0x7f45980156a8 "test", new_name=0x0, create_info=0x7f45dc5ef088, table_list=0x7f4598015070, alter_info=0x7f45dc5eefe0, order_num=0, order=0x0, ignore=false) at sql_table.cc:9666
#14 Sql_cmd_alter_table::execute (this=0x7f45980156b8, thd=0x7f4598000d50) at sql_alter.cc:325
#15 mysql_execute_command (thd=0x7f4598000d50) at sql_parse.cc:6258
#16 mysql_parse (thd=0x7f4598000d50, rawbuf=0x7f4598014f58 "alter table t drop system versioning, algorithm=inplace", length=55, parser_state=0x7f45dc5f2410, is_com_multi=false, is_next_command=false) at sql_parse.cc:7988
```

File: sql/handler.cc [hunk 2]
```diff
@@ -7374,38 +7375,74 @@ bool Vers_parse_info::check_with_conditions(const char *table_name) const
...
-bool Vers_parse_info::check_generated_type(const char *table_name,
-                                           Alter_info *alter_info,
-                                           bool integer_fields) const
+bool Vers_parse_info::check_sys_fields(const char *table_name,
+                                       Alter_info *alter_info,
+                                       bool native) const
 {
   List_iterator<Create_field> it(alter_info->create_list);
+  vers_sys_type_t found= VERS_UNDEFINED;
+  uint found_flag= 0;
```
Comment (kevgs) [inline sql/handler.cc:7384]: Make variable `const`.

File: sql/handler.cc [hunk 3]
```diff
@@ -7374,38 +7375,74 @@ bool Vers_parse_info::check_with_conditions(const char *table_name) const
...
   while (Create_field *f= it++)
   {
-    if (is_start(*f) || is_end(*f))
+    vers_sys_type_t check_unit= VERS_UNDEFINED;
+    uint sys_flag= f->flags & VERS_SYSTEM_FIELD;
+
+    if (!sys_flag)
+      continue;
+
+    if (sys_flag & found_flag)
```
Comment (kevgs) [inline sql/handler.cc:7393]: Always `falste`. And BTW it's already checked in `sql_yacc.yy`.

File: sql/handler.cc [hunk 4]
```diff
@@ -7374,38 +7375,74 @@ bool Vers_parse_info::check_with_conditions(const char *table_name) const
...
+                    "BIGINT(20) UNSIGNED",
+                  table_name);
+        return true;
       }
+      found= check_unit;
     }
   }
 
-  return false;
+  my_error(ER_MISSING, MYF(0), table_name, found_flag & VERS_SYS_START_FLAG ?
```
Comment (kevgs) [inline sql/handler.cc:7447]: Always will be "ROW START/END".

File: storage/innobase/trx/trx0trx.cc (hunk 1 above)
Comment (midenok) [reply storage/innobase/trx/trx0trx.cc:72]: This is storage format, not machine format.

File: sql/handler.cc (hunk 3 above)
Comment (midenok) [reply sql/handler.cc:7393]: I know. This check is natural here. Nothing wrong with double check.

File: sql/handler.cc (hunk 2 above)
Comment (midenok) [reply sql/handler.cc:7384]: Typo fixed.

File: storage/innobase/include/trx0trx.h [hunk 5]
```diff
@@ -786,6 +786,7 @@ class trx_mod_table_time_t
 	undo_no_t	first;
 	/** First modification of a system versioned column */
 	undo_no_t	first_versioned;
+	bool		vers_by_trx;
```
Comment (kevgs) [inline storage/innobase/include/trx0trx.h:789]: Comment is missing.

File: storage/innobase/include/dict0types.h [hunk 6]
```diff
@@ -57,6 +57,7 @@ typedef ib_id_t		index_id_t;
 
 /** The bit pattern corresponding to TRX_ID_MAX */
 extern const char trx_id_max_bytes[8];
+extern const char timestamp_max_bytes[7];
```
Comment (kevgs) [inline storage/innobase/include/dict0types.h:60]: Comment is missing.

File: sql/handler.cc [hunk 7]
```diff
@@ -7374,38 +7375,76 @@ bool Vers_parse_info::check_with_conditions(const char *table_name) const
...
-          return true;
-        }
-      }
-      else
+      my_error(ER_VERS_DUPLICATE_ROW_START_END, MYF(0),
+                found_flag & VERS_SYS_START_FLAG ? "START" : "END", f->field_name.str);
+      return true;
+    }
+
+    sys_flag|= found_flag;
```
Comment (kevgs) [inline sql/handler.cc:7400]: `found_flag|= sys_flag`?

**Review by kevgs — APPROVED** [review]

**Pull request merged**